lib/ai/model/tool.go

```diff
@@ -109,18 +134,281 @@ func (*commandExecutionTool) parseInput(input string) (*commandExecutionToolInpu
 	return &output, nil
 }
 
-type NodeGetter interface {
-	GetNodes(ctx context.Context, fn func(n services.Node) bool) []types.Server
-	NodeCount() int
+type accessRequestListRequestableRolesTool struct{}
+
+func (*accessRequestListRequestableRolesTool) Name() string {
+	return "List Requestable Roles"
+}
+
+func (*accessRequestListRequestableRolesTool) Description() string {
+	return "List all roles that can be requested via access requests."
+}
+
+func (a *accessRequestListRequestableRolesTool) Run(ctx context.Context, toolCtx *ToolContext, input string) (string, error) {
+	roles := toolCtx.AccessChecker.Roles()
+	requestable := make(map[string]struct{}, 0)
+	for _, role := range roles {
+		for _, requestableRole := range role.GetAccessRequestConditions(types.Allow).Roles {
+			requestable[requestableRole] = struct{}{}
+		}
+	}
+	for _, role := range roles {
+		for _, requestableRole := range role.GetAccessRequestConditions(types.Deny).Roles {
+			delete(requestable, requestableRole)
+		}
+	}
+
+	resp := strings.Builder{}
+	for role := range requestable {
+		resp.Write([]byte(role))
+		resp.Write([]byte("\n"))
+	}
+
+	if resp.Len() == 0 {
+		return "No requestable roles found", nil
+	}
+
+	return resp.String(), nil
+}
+
+type accessRequestListRequestableResourcesTool struct{}
+
+func (*accessRequestListRequestableResourcesTool) Name() string {
+	return "List Requestable Resources"
 }
 
-type embeddingRetrievalTool struct {
-	assistClient      assist.AssistEmbeddingServiceClient
-	nodeClient        NodeGetter
-	userAccessChecker services.AccessChecker
-	currentUser       string
+func (*accessRequestListRequestableResourcesTool) Description() string {
+	return `List all resources with IDs that can be requested via access requests.
+This includes nodes via SSH access.`
 }
 
+func (a *accessRequestListRequestableResourcesTool) Run(ctx context.Context, toolCtx *ToolContext, input string) (string, error) {
+	foundResources := make(chan promptResource, 0)
+	workersAlive := new(int32)
+	*workersAlive = 5
+
+	searchAndAppend := func(resourceType string, convert func(types.Resource) (promptResource, error)) error {
+		defer func() {
+			if atomic.AddInt32(workersAlive, -1) == 0 {
+				close(foundResources)
+			}
+		}()
+
+		list, err := toolCtx.ListResources(ctx, proto.ListResourcesRequest{
```

Comment:
Are we making calls to Auth, or we're reading from the Proxy cache?

Reply:
Let's leave it for now as it is. If it causes problems, we can move that logic to the auth (similar to what we did with embeddings) or find a different alternative.

Reply:
👍